torch/nn/functional.py

```diff
@@ -1395,15 +1395,7 @@ def linear(input, weight, bias=None):
         - Bias: :math:`(out\_features)`
         - Output: :math:`(N, *, out\_features)`
     """
-    if input.dim() == 2 and bias is not None:
-        # fused op is marginally faster
-        ret = torch.addmm(bias, input, weight.t())
-    else:
-        output = input.matmul(weight.t())
-        if bias is not None:
-            output += bias
-        ret = output
-    return ret
+    return torch._C._nn.linear(input, weight, bias)
```

Comment:
I wonder whether it'd break some of the jit passes as now models would have a lot of linears. @zdevito @apaszke @ZolotukhinM 

As Greg said it might be safer/easier to just implement 2-d version of matmul and addmm for mkldnn and call it a day. Or is your worry that `.t()` will be materialized explicitly?

we could also just put dispatch in python here (i.e. `if input.__is_mkldnn() ...`)

Reply:
I'm actually not concerned about the materialized transpose here because linear only takes negligible amount of the computation time in resnext :-).
It's just when I look at the profiler or the tracer, it's addmm instead of linear showing up (and sometimes it's matmul) annoys me. For now I can live with that to unblock the land.